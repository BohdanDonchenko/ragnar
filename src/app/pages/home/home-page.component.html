<h2>Home</h2>

<data-component title="Update" [data]="data$ | async" (onUpdated)="update()">
</data-component>
<data-component title="Server update" [data]="serverData$ | async" (onUpdated)="serverUpdate()">
</data-component>
<div>
  Filtered Data: {{serverFilteredData$ | async}}
</div>
