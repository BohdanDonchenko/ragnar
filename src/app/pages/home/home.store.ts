import { BehaviorSubject } from 'rxjs';
import { Injectable } from '@angular/core';

@Injectable()
export class HomeStore {
  readonly counter$ = new BehaviorSubject<number | null>(null);

  readonly serverCounter$ = new BehaviorSubject<number>(0);
}
