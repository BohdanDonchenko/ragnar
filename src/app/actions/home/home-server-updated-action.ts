import { Injectable } from '@angular/core';
import { BaseAction } from 'app/actions/base-action';
import { ValuesRepository } from 'app/repositories/values-repository';
import { Store } from 'app/store/store';
import { StoreAccessor } from 'app/store/store-accessor';

@Injectable()
export class HomeServerUpdatedAction extends BaseAction {
  constructor(
    storeAccessor: StoreAccessor,
    private valuesRepository: ValuesRepository
  ) {
    super(storeAccessor);
  }

  protected async callback(store: Store): Promise<Store> {
    store.homeStore.serverCounter = await this.valuesRepository.getData();
    return store;
  }
}
