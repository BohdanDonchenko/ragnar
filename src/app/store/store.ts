import { HomeStore } from 'app/store/home/home-store';

export class Store {
  private _homeStore = new HomeStore();
  
  get homeStore(): HomeStore {
    return this._homeStore;
  }
}
