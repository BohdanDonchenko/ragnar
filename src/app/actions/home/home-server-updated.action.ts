import { IAction } from '@actions/i-action';
import { Injectable } from '@angular/core';
import { ValuesRepository } from '@repositories/values.repository';
import { HomeStore } from '@stores/home.store';

@Injectable()
export class HomeServerUpdatedAction implements IAction {
  constructor(
    private store: HomeStore,
    private valuesRepository: ValuesRepository
  ) {}

  execute() {
    this.valuesRepository
      .getData()
      .subscribe(value => this.store.serverCounter$.next(value));
  }
}
