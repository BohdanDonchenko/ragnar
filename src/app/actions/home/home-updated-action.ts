import { Injectable } from '@angular/core';
import { BaseAction } from 'app/actions/base-action';
import { Store } from 'app/store/store';
import { StoreAccessor } from 'app/store/store-accessor';

@Injectable()
export class HomeUpdatedAction extends BaseAction {
  constructor(storeAccessor: StoreAccessor) {
    super(storeAccessor);
  }

  protected callback(store: Store): Store | Promise<Store> {
    store.counter++;
    return store;
  }
}
