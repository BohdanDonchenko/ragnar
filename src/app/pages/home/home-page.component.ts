import { Component } from '@angular/core';
import { Observable } from 'rxjs';
import { filter } from 'rxjs/operators';
import { HomeStore } from './home.store';
import { ServerUpdatedCommand } from './commands/server-updated.command';
import { UpdatedCommand } from './commands/updated.command';

@Component({
  selector: 'home-page',
  templateUrl: 'home-page.component.html',
})
export class HomePageComponent {
  data$: Observable<number | null>;

  serverData$: Observable<number>;

  serverFilteredData$: Observable<number>;

  constructor(
    store: HomeStore,
    private homeUpdatedCommand: UpdatedCommand,
    private homeServerUpdatedCommand: ServerUpdatedCommand,
  ) {
    this.data$ = store.counter$;
    this.serverData$ = store.serverCounter$;
    this.serverFilteredData$ = store.serverCounter$.pipe(
      filter(v => v % 3 === 0),
    );
  }

  update() {
    this.homeUpdatedCommand.execute();
  }

  serverUpdate() {
    this.homeServerUpdatedCommand.execute();
  }
}
