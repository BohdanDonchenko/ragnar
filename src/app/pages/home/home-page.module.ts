import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ServerUpdatedCommand } from './commands/server-updated.command';
import { HomeRoutingModule } from './home-routing.module';

import { HomePageComponent } from './home-page.component';
import { DataComponent } from './components/home-data.component/data.component';
import { HomeStore } from "./home.store";
import { UpdatedCommand } from './commands/updated.command';

@NgModule({
  imports: [CommonModule, HomeRoutingModule],
  exports: [HomePageComponent],
  declarations: [HomePageComponent, DataComponent],
  providers: [HomeStore, UpdatedCommand, ServerUpdatedCommand]
})
export class HomePageModule { }
