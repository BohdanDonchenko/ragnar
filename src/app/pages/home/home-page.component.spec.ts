import { TestBed, async } from '@angular/core/testing';
import { take } from 'rxjs/operators';
import { RepositoriesModule } from '../../repositories/repositories.module';
import { ServicesModule } from '../../services/services.module';
import { ServerUpdatedCommand } from './commands/server-updated.command';
import { UpdatedCommand } from './commands/updated.command';
import { HomeStore } from './home.store';
import { HomePageComponent } from './home-page.component';
import { DataComponent } from './components/home-data.component/data.component';

describe('HomePageComponent', () => {
  describe('Snapshot', () => {
    beforeEach(async(() => {
      TestBed.configureTestingModule({
        imports: [ServicesModule, RepositoriesModule],
        providers: [HomeStore, UpdatedCommand, ServerUpdatedCommand],
        declarations: [HomePageComponent, DataComponent],
      }).compileComponents();
    }));

    test('check snapshot', () => {
      const fixture = TestBed.createComponent(HomePageComponent);

      expect(fixture).toMatchSnapshot();
    });
  });

  describe('Functionality', () => {
    let store = new HomeStore();
    let updCommand = {} as UpdatedCommand;
    let serverUpdCommand = {} as ServerUpdatedCommand;

    beforeEach(() => {
      store = new HomeStore();
      updCommand = {} as UpdatedCommand;
      serverUpdCommand = {} as ServerUpdatedCommand;
    });

    test('constructor', () => {
      const sut = new HomePageComponent(store, updCommand, serverUpdCommand);
      store.counter$.next(1);
      sut.data$.pipe(take(1)).subscribe(v => expect(v).toBe(1));

      store.serverCounter$.next(1);
      sut.serverData$.pipe(take(1)).subscribe(v => expect(v).toBe(1));

      store.counter$.next(9);
      sut.serverFilteredData$.pipe(take(1)).subscribe(v => expect(v).toBe(9));
    });

    test('update', () => {
      updCommand.execute = jest.fn();
      const sut = new HomePageComponent(store, updCommand, serverUpdCommand);
      sut.update();
      expect(updCommand.execute).toBeCalledTimes(1);
    });

    test('serverUpdate', () => {
      serverUpdCommand.execute = jest.fn();
      const sut = new HomePageComponent(store, updCommand, serverUpdCommand);
      sut.serverUpdate();
      expect(serverUpdCommand.execute).toBeCalledTimes(1);
    });
  });
});
