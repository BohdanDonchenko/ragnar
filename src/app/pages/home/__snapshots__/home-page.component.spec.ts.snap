// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePageComponent Snapshot check snapshot 1`] = `
<home-page
  data$={[Function BehaviorSubject]}
  homeServerUpdatedCommand={[Function ServerUpdatedCommand]}
  homeUpdatedCommand={[Function UpdatedCommand]}
  serverData$={[Function BehaviorSubject]}
  serverFilteredData$={[Function AnonymousSubject]}
>
  <h2>
    Home
  </h2><data-component
    title="Update"
  >
    <div
      class="home-data"
    >
      <button>
        
      </button>
      <h4>
        Data: 
      </h4>
    </div>
  </data-component><data-component
    title="Server update"
  >
    <div
      class="home-data"
    >
      <button>
        
      </button>
      <h4>
        Data: 
      </h4>
    </div>
  </data-component><div>
     Filtered Data: 
  </div>
</home-page>
`;
